### DISCUSSION TITLE
Dev UI Ideas

### DISCUSSION THREAD

**Discussion by phillip-kruger** [quarkusio/quarkus#35178] [Community]
With the new Dev UI ported over and all features from the old Dev UI available, I want to keep a list of ideas for new features for the Dev UI. We can then prioritize those and start implementing them. It is a living list, so please add to it at any time and comment on existing list item ideas to increase their priority

**Dev UI Ideas** 

- [ ] Show when a new version of Quarkus is available
- [ ] Bundle all front end resources as as part on the build 
- [ ] Show a pipeline of message passed in/out from a channel for ReactiveMessaging
- [ ] Hide/Show transport secrets so we could have it enabled when running devmode remotely/in cluster
- [ ] Direct Zulip User stream integration in Dev UI
- [x] Create a way to share components and other Dev UI related things between extension or from a common module/extension (#44275)
- [ ] Find a way do allow certain parts in a PROD environment.
- [x] Add localization (see https://lit.dev/docs/localization/overview/)
- [ ] Have a way to select and run only a specific test instead of run all or all failed ones
- [ ] Replace code editor with something like https://microsoft.github.io/monaco-editor/
- [x] Add a way to browse all available extensions and install straight from Dev UI. Similarly allow removing an extension from Dev UI. (#43840)
- [x] Show project README in Dev UI
- [x] Replace the static page on root (/) with a dynamic page created by Dev UI (#37680)
- [x] Upgrade to codemirror 6 (#38578)
- [x] Add key binding (#39459)
- [x] Page that show all known endpoints (like the 404 page) (#37039)

**Comment by phillip-kruger** [comment]
Add a way to browse all available extensions and install straight from Dev UI. Similarly allow removing an extension from Dev UI

**Comment by phillip-kruger** [comment]
Show when a new version of Quarkus is available

**Comment by phillip-kruger** [comment]
Page that show all known endpoints (like the 404 page)

**Comment by phillip-kruger** [comment]
Find a way do allow certain parts in a PROD environment.

**Comment by joes** [comment]
I run apps during development in kubernetes and indicate to the app that it should run in dev mode with the environment variable:

```yaml
            - name: QUARKUS_LAUNCH_DEVMODE
              value: "true"
```

But, alas, the dev-ui is not available in the app.

Would it be possible to make the dev-ui available if the app is run in devmode?

**Reply by phillip-kruger** [reply]
Not sure if this is possible (Did this work with the old Dev UI? ), but "Find a way do allow certain parts in a PROD environment." might be enough for you. What features of Dev UI are you after ?

**Reply by josephearl** [reply]
The Swagger/Open API endpoints page would be very useful, especially in pre-prod environments for testing

**Reply by gsmet** [reply]
@josephearl you can already expose Swagger UI with [`quarkus.swagger-ui.always-include`](https://quarkus.io/guides/openapi-swaggerui#quarkus-swaggerui_quarkus.swagger-ui.always-include).

**Reply by josephearl** [reply]
Awesome thanks!

**Reply by hantsy** [reply]
It provides visual UI to show a pipeline of message passed in/out from a channel when ReactiveMessaging extension is enabled?

**Reply by maxandersen** [reply]
How about the feature to enable Devui to hide/not transport secrets so we could have it enabled when running devmode remotely/in cluster?

**Reply by joes** [reply]
@phillip-kruger 

This did not work with the old dev ui. I guess I am looking for the total package if that is at all possible. Is there some technical limitation that makes this impossible? I build with quarkus.profile=dev and set the launch mode as shown before. Why would this not build the application with the dev ui enabled?

**Reply by phillip-kruger** [reply]
@joes ok - Yea I am not sure, we will have to investigate. You are welcome to see if you can find out why it does not work (i.e if there is a technical reason).

**Comment by phillip-kruger** [comment]
Bundle all front end resources as as part on the build

**Comment by phillip-kruger** [comment]
Replace the static page on root (/) with a dynamic page created by Dev UI

**Reply by maxandersen** [reply]
That doesn't sound like a good idea. 

Devui should stay under one sub path (/q/ by default) for security reasons and we need to guarantee the first page loads no matter what - especially when running the app where devui is not present.

**Reply by phillip-kruger** [reply]
Yes, it will. The idea is to redirect to /q/dev-ui/welcome if there is not user static file

**Reply by phillip-kruger** [reply]
So when going to root (/) we usually service the static file (that can be replaced with a users own content). We now replace that static file with a Route that (if no user index.html exist) redirect to /q/dev-ui/welcome that serve the page (now dynamic and up to date)

**Reply by maxandersen** [reply]
Yes - but then you get that there is no index page when you run without devmode..?

**Reply by phillip-kruger** [reply]
I think that is the point. That static page should not be available outside dev mode. Currently you have to delete the generated static page, else you have that in PROD mode, that you don't want.

**Reply by maxandersen** [reply]
the advantage of the page being there is that it can be easily updated rather than be surprised that outside devmode it is missing.

That said having a more adaptive dev default front page - one that even can have a "generate index.html in the right place" button is intriguing.

we should ensure it so that route is lowest priority compared to any routes the user have installed.

**Comment by cvgaviao** [comment]
I would like to have a way to select and run only a specific test instead of run all or all failed ones.
